File: lib/domain/fertilize/entities/fertilize_entity.rb
# frozen_string_literal: true

module Domain
  module Fertilize
    module Entities
      class FertilizeEntity
        attr_reader :id, :user_id, :name, :n, :p, :k, :description, :package_size, :region, :is_reference, :created_at, :updated_at

        def initialize(attributes)
          @id = attributes[:id]
          @user_id = attributes[:user_id]
          @name = attributes[:name]
          @n = attributes[:n]
          @p = attributes[:p]
          @k = attributes[:k]
          @description = attributes[:description]
          @package_size = attributes[:package_size]
          @region = attributes[:region]
          @is_reference = attributes[:is_reference]
          @created_at = attributes[:created_at]
          @updated_at = attributes[:updated_at]

          validate!
        end

        def to_model
          ::Fertilize.find(id)
        end

        def self.from_model(record)
          new(
            id: record.id,
            user_id: record.user_id,
            name: record.name,
            n: record.n,
            p: record.p,
            k: record.k,
            description: record.description,
            package_size: record.package_size,
            region: record.region,
            is_reference: record.is_reference,
            created_at: record.created_at,
            updated_at: record.updated_at
          )
        end
        
        def reference?
          !!is_reference
        end
        
        def has_nutrient?(nutrient)
          case nutrient.to_sym
          when :n
            n.present? && n > 0
          when :p
            p.present? && p > 0
          when :k
            k.present? && k > 0
          else
            false
          end
        end
        
        def npk_summary
          [n, p, k].compact.map { |v| v.to_i }.join('-')
        end
        
        private
        
        def validate!
          raise ArgumentError, "Name is required" if name.blank?
        end
      end
    end
  end
end

File: crates/agrr-domain/src/fertilize/entities/fertilize_entity.rs
use crate::shared::hash::blank_attr;
use crate::shared::record_ref::RecordRef;

/// Ruby: `Domain::Fertilize::Entities::FertilizeEntity`
#[derive(Debug, Clone, PartialEq)]
pub struct FertilizeEntity {
    pub id: Option<i64>,
    pub user_id: Option<i64>,
    pub name: String,
    pub n: Option<f64>,
    pub p: Option<f64>,
    pub k: Option<f64>,
    pub description: Option<String>,
    pub package_size: Option<f64>,
    pub region: Option<String>,
    pub is_reference: bool,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct FertilizeEntityAttrs {
    pub id: Option<i64>,
    pub user_id: Option<i64>,
    pub name: String,
    pub n: Option<f64>,
    pub p: Option<f64>,
    pub k: Option<f64>,
    pub description: Option<String>,
    pub package_size: Option<f64>,
    pub region: Option<String>,
    pub is_reference: bool,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
}

impl FertilizeEntity {
    pub fn new(attrs: FertilizeEntityAttrs) -> Result<Self, String> {
        let entity = Self {
            id: attrs.id,
            user_id: attrs.user_id,
            name: attrs.name,
            n: attrs.n,
            p: attrs.p,
            k: attrs.k,
            description: attrs.description,
            package_size: attrs.package_size,
            region: attrs.region,
            is_reference: attrs.is_reference,
            created_at: attrs.created_at,
            updated_at: attrs.updated_at,
        };
        entity.validate()?;
        Ok(entity)
    }

    pub fn reference(&self) -> bool {
        self.is_reference
    }

    pub fn has_nutrient(&self, nutrient: &str) -> bool {
        match nutrient {
            "n" => self.n.map(|v| v > 0.0).unwrap_or(false),
            "p" => self.p.map(|v| v > 0.0).unwrap_or(false),
            "k" => self.k.map(|v| v > 0.0).unwrap_or(false),
            _ => false,
        }
    }

    pub fn npk_summary(&self) -> String {
        [self.n, self.p, self.k]
            .into_iter()
            .flatten()
            .map(|v| v.trunc() as i64)
            .map(|v| v.to_string())
            .collect::<Vec<_>>()
            .join("-")
    }

    fn validate(&self) -> Result<(), String> {
        if blank_attr(&crate::shared::attr::AttrValue::Str(self.name.clone())) {
            return Err("Name is required".into());
        }
        Ok(())
    }
}

impl RecordRef for FertilizeEntity {
    fn is_reference(&self) -> bool {
        self.is_reference
    }

    fn user_id(&self) -> Option<i64> {
        self.user_id
    }
}

#[cfg(test)]
mod entities_fertilize_entity_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/fertilize/entities_fertilize_entity_test.rs"));
}
